In [1]:
import pandas as pd

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [55]:
df = pd.read_excel("C:\\Users\\aksha\\Desktop\\ng2d3\\Online Retail.xlsx")
df.head()
df1 = df

In [45]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [46]:
df1.Country.nunique()

38

there are 38 unique countrie as follows:

In [47]:
df1.Country.unique()

array([u'United Kingdom', u'France', u'Australia', u'Netherlands',
       u'Germany', u'Norway', u'EIRE', u'Switzerland', u'Spain', u'Poland',
       u'Portugal', u'Italy', u'Belgium', u'Lithuania', u'Japan',
       u'Iceland', u'Channel Islands', u'Denmark', u'Cyprus', u'Sweden',
       u'Austria', u'Israel', u'Finland', u'Bahrain', u'Greece',
       u'Hong Kong', u'Singapore', u'Lebanon', u'United Arab Emirates',
       u'Saudi Arabia', u'Czech Republic', u'Canada', u'Unspecified',
       u'Brazil', u'USA', u'European Community', u'Malta', u'RSA'], dtype=object)

<h4>Drop rows with same customerID and country</h4>

In [48]:
customer_country=df1[['Country','CustomerID']].drop_duplicates()

In [49]:
customer_country.head()

Country  CustomerID
0   United Kingdom     17850.0
9   United Kingdom     13047.0
26          France     12583.0
46  United Kingdom     13748.0
65  United Kingdom     15100.0

<h4>First Group by country and aggregate on Cutomer ID to find how many customers from a particular country then sort by customer count</h4> 

In [50]:
customer_country.groupby(['Country'])['CustomerID'].aggregate('count').reset_index().sort_values('CustomerID', ascending=False)

Country  CustomerID
36        United Kingdom        3950
14               Germany          95
13                France          87
31                 Spain          31
3                Belgium          25
33           Switzerland          21
27              Portugal          19
19                 Italy          15
12               Finland          12
1                Austria          11
25                Norway          10
24           Netherlands           9
0              Australia           9
6        Channel Islands           9
9                Denmark           9
7                 Cyprus           8
32                Sweden           8
20                 Japan           8
26                Poland           6
34                   USA           4
5                 Canada           4
37           Unspecified           4
18                Israel           4
15                Greece           4
10                  EIRE           3
23                 Malta           2
35  United Arab Emirates           2
2                Bahrain           2
22             Lithuania           1
8         Czech Republic           1
21               Lebanon           1
28                   RSA           1
29          Saudi Arabia           1
30             Singapore           1
17               Iceland           1
4                 Brazil           1
11    European Community           1
16             Hong Kong           0

As more than 90% of the customers are from UK we will select only that country

In [66]:
df1.loc[df1['Country']=='United Kingdom'].head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

Check for null values in each column


In [68]:
df1.isnull().sum(axis=0)

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

We will remove these null values 

In [69]:
df1=df1[pd.notnull(df1['CustomerID'])]

In [70]:
df1.isnull().sum(axis=0)

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [71]:
df1.Quantity.min()

-80995

remove the negative values

In [73]:
df1=df1[df['Quantity']>0]

In [75]:
df1.shape

(397924, 8)

In [76]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 397924 entries, 0 to 541908
Data columns (total 8 columns):
InvoiceNo      397924 non-null object
StockCode      397924 non-null object
Description    397924 non-null object
Quantity       397924 non-null int64
InvoiceDate    397924 non-null datetime64[ns]
UnitPrice      397924 non-null float64
CustomerID     397924 non-null float64
Country        397924 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.3+ MB


We are dealing with 397924 entries and 8 columns 



<h4>Add a column for total price</h4>

In [77]:
df1['Total Price']=df1['UnitPrice']*df1['Quantity']

In [78]:
df1.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Total Price  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom        15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom        22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34

<h4>First and last order date in data</h4>

In [79]:
df1.InvoiceDate.min()

Timestamp('2010-12-01 08:26:00')

In [80]:
df1.InvoiceDate.max()

Timestamp('2011-12-09 12:50:00')

<h3>For recency we will take 2011-12-10 </h3>

In [159]:
import datetime as dt
NOW = dt.datetime(2011,12,10)

Convert Date string into date object

In [160]:
df1['InvoiceDate']=pd.to_datetime(df['InvoiceDate'])

In [161]:
rfmTable=df1.groupby('CustomerID').agg({'InvoiceDate':lambda x:(NOW-x.max()).days,'InvoiceNo':lambda x:len(x),'Total Price':lambda x: x.sum()})

In [162]:
rfmTable.head()

InvoiceNo  Total Price  InvoiceDate
CustomerID                                     
12346.0             1     77183.60          325
12347.0           182      4310.00            2
12348.0            31      1797.24           75
12349.0            73      1757.55           18
12350.0            17       334.40          310

In [163]:
rfmTable['InvoiceDate']=rfmTable['InvoiceDate'].astype(int)

In [164]:
rfmTable.rename(columns={'InvoiceDate':'Recency','InvoiceNo':'Frequency','Total Price':'monetory_value'},inplace=True)

In [165]:
rfmTable.head()

Frequency  monetory_value  Recency
CustomerID                                    
12346.0             1        77183.60      325
12347.0           182         4310.00        2
12348.0            31         1797.24       75
12349.0            73         1757.55       18
12350.0            17          334.40      310

<h3>Customer ID 12346 has purchsed once 325 days from last entry date and its monetory value was 77183.60</h3>

Its obvious that 1st customer is not our best customer

In [166]:
quantiles = rfmTable.quantile(q=[0.25,0.5,0.75])

In [167]:
quantiles

Frequency  monetory_value  Recency
0.25       17.0         307.245     17.0
0.50       41.0         674.450     50.0
0.75      100.0        1661.640    141.5

So we have 25% customers who bought less than 17 times, 50%customers who bought less than 41 times and 70% who bought less than 100 times.


In [168]:
rfmTable.describe()

Frequency  monetory_value      Recency
count  4339.000000     4339.000000  4339.000000
mean     91.708689     2053.793018    92.041484
std     228.792852     8988.248381   100.007757
min       1.000000        0.000000     0.000000
25%      17.000000      307.245000    17.000000
50%      41.000000      674.450000    50.000000
75%     100.000000     1661.640000   141.500000
max    7847.000000   280206.020000   373.000000

In [169]:
quantiles=quantiles.to_dict()

Lowest recency, highest frequency and monetary are our best customers.

In [170]:
segmented_rfm = rfmTable
segmented_rfm['Recency'].head()

CustomerID
12346.0    325
12347.0      2
12348.0     75
12349.0     18
12350.0    310
Name: Recency, dtype: int32

In [171]:
def RScore(x,p,d):
    if x<=d[p][0.25]:
        return 4
    elif x<=d[p][0.5]:
        return 3
    elif x<=d[p][0.75]:
        return 2
    else:
        return 1
def FMScore(x,p,d):
    if x<=d[p][0.25]:
        return 1
    elif x<=d[p][0.5]:
        return 2
    elif x<=d[p][0.75]:
        return 3
    else:
        return 4

In [172]:
segmented_rfm['r_quartile'] = segmented_rfm['Recency'].apply(RScore, args=('Recency',quantiles,))
segmented_rfm['f_quartile'] = segmented_rfm['Frequency'].apply(FMScore, args=('Frequency',quantiles,))
segmented_rfm['m_quartile'] = segmented_rfm['monetory_value'].apply(FMScore, args=('monetory_value',quantiles,))

In [173]:
segmented_rfm.head()

Frequency  monetory_value  Recency  r_quartile  f_quartile  \
CustomerID                                                               
12346.0             1        77183.60      325           1           1   
12347.0           182         4310.00        2           4           4   
12348.0            31         1797.24       75           2           2   
12349.0            73         1757.55       18           3           3   
12350.0            17          334.40      310           1           1   

            m_quartile  
CustomerID              
12346.0              4  
12347.0              4  
12348.0              4  
12349.0              4  
12350.0              2

In [174]:
segmented_rfm['RFMScore']= segmented_rfm.r_quartile.map(str)+segmented_rfm.f_quartile.map(str)+segmented_rfm.m_quartile.map(str)

In [175]:
segmented_rfm.head()

Frequency  monetory_value  Recency  r_quartile  f_quartile  \
CustomerID                                                               
12346.0             1        77183.60      325           1           1   
12347.0           182         4310.00        2           4           4   
12348.0            31         1797.24       75           2           2   
12349.0            73         1757.55       18           3           3   
12350.0            17          334.40      310           1           1   

            m_quartile RFMScore  
CustomerID                       
12346.0              4      114  
12347.0              4      444  
12348.0              4      224  
12349.0              4      334  
12350.0              2      112

In [181]:
segmented_rfm[segmented_rfm['RFMScore']=='444'].sort_values('monetory_value',ascending=False).head(10)

Frequency  monetory_value  Recency  r_quartile  f_quartile  \
CustomerID                                                               
14646.0          2080       280206.02        1           4           4   
18102.0           431       259657.30        0           4           4   
17450.0           337       194550.79        8           4           4   
14911.0          5677       143825.06        1           4           4   
14156.0          1400       117379.63        9           4           4   
17511.0           963        91062.38        2           4           4   
16684.0           277        66653.56        4           4           4   
14096.0          5111        65164.79        4           4           4   
13694.0           568        65039.62        3           4           4   
15311.0          2379        60767.90        0           4           4   

            m_quartile RFMScore  
CustomerID                       
14646.0              4      444  
18102.0              4      444  
17450.0              4      444  
14911.0              4      444  
14156.0              4      444  
17511.0              4      444  
16684.0              4      444  
14096.0              4      444  
13694.0              4      444  
15311.0              4      444